File: python/selection_sort.py
def find_smallest(arr):
  smallest_value = arr[0]
  smallest_index = 0

  for i in range(1, len(arr)):
    if arr[i] < smallest_value:
      smallest_value = arr[i]
      smallest_index = i
  
  return smallest_index

# return a sorted list
# selection sort will insert the results in a new array
def selection_sort(arr):
  result = []

  # not looping through each item actually, 
  # but instead setting up a loop that will loop 
  # for the same number of items in the array
  for i in range(1, len(arr)): 
    smallest_index = find_smallest(arr)
    result.append(arr.pop(smallest_index))

  return result

print(selection_sort([5, 3, 6, 2, 100]))

File: elixir/selection_sort.exs
defmodule SelectionSort do
  defp find_smallest_index(tuple, i, n, smallest_index, smallest_value) when i < n do
    case elem(tuple, i) do
      x when x < smallest_value ->
        find_smallest_index(tuple, i + 1, tuple_size(tuple), i, x)
      _ ->
        find_smallest_index(tuple, i + 1, tuple_size(tuple), smallest_index, smallest_value)
    end
  end

  defp find_smallest_index(_tuple, _i, _n, smallest_index, _smallest_value) do
    smallest_index
  end

  def sort(tuple, result) when tuple_size(tuple) > 0 do
    smallest_index = find_smallest_index(tuple, 0, tuple_size(tuple), 0, elem(tuple, 0))
    smallest_value = elem(tuple, smallest_index)
    result = result ++ [smallest_value]
    tuple = Tuple.delete_at(tuple, smallest_index)
    sort(tuple, result)
  end

  def sort(_tuple, result) do
    result
  end
end

tuple = {2, 5, 6, 1, 9, 3, 8, 7}
IO.inspect(SelectionSort.sort(tuple, []))
